### ISSUE TITLE
`Linear_Scale_To_Linear`: Out of bounds access of `remap` buffer on GDI Mission 9

### ISSUE THREAD

**Issue by giulianobelinassi** [TheAssemblyArmada/Vanilla-Conquer#658]
On GDI Mission 9, with `-fsanitize=address`, the game crashes due to Out of Bounds access when radar is enabled.

It is hard to track what this function does, as it looks like assembler code compiled to C by some disassembler.
What should it do?

==1120727==ERROR: AddressSanitizer: global-buffer-overflow on address 0x55555596eacd at pc 0x5555558630a9 bp 0x7fffffffd570 sp 0x7fffffffd568
READ of size 1 at 0x55555596eacd thread T0
    #0 0x5555558630a8 in Linear_Scale_To_Linear(void*, void*, int, int, int, int, int, int, int, int, bool, char*) /home/giulianob/bagunca/Vanilla-Conquer/common/linear.cpp:196
    #1 0x5555558510b7 in GraphicViewPortClass::Scale(GraphicViewPortClass&, int, int, int, int, int, int, int, int, bool, char*) /home/giulianob/bagunca/Vanilla-Conquer/common/graphicsviewport.cpp:664
    #2 0x55555577c596 in RadarClass::Render_Terrain(short, int, int, int) /home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/radar.cpp:559
    #3 0x55555577e31f in RadarClass::Plot_Radar_Pixel(short) /home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/radar.cpp:852
    #4 0x55555577bab9 in RadarClass::Draw_It(bool) /home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/radar.cpp:449
    #5 0x5555557683d0 in PowerClass::Draw_It(bool) /home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/power.cpp:249
    #6 0x5555557ace5b in SidebarClass::Draw_It(bool) /home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/sidebar.cpp:696
    #7 0x5555557c3512 in TabClass::Draw_It(bool) /home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/tab.cpp:84
    #8 0x5555556b8ada in HelpClass::Draw_It(bool) /home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/help.cpp:263
    #9 0x5555556b6077 in GScreenClass::Render() /home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/gscreen.cpp:390
    #10 0x55555564037b in Main_Loop() (/home/giulianob/win/cnc95_orig/vanillatd+0xec37b)
    #11 0x55555563b9bd in Main_Game(int, char**) (/home/giulianob/win/cnc95_orig/vanillatd+0xe79bd)
    #12 0x5555557c12c1 in main /home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/startup.cpp:376
    #13 0x7ffff6fded09 in __libc_start_main ../csu/libc-start.c:308
    #14 0x5555555d6e49 in _start (/home/giulianob/win/cnc95_orig/vanillatd+0x82e49)

0x55555596eacd is located 13 bytes to the right of global variable 'Layer' defined in '/home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/display.cpp:86:12' (0x55555596ea60) of size 96
0x55555596eacd is located 19 bytes to the left of global variable 'FadingBrighten' defined in '/home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/display.cpp:91:15' (0x55555596eae0) of size 256
SUMMARY: AddressSanitizer: global-buffer-overflow /home/giulianob/bagunca/Vanilla-Conquer/common/linear.cpp:196 in Linear_Scale_To_Linear(void*, void*, int, int, int, int, int, int, int, int, bool, char*)
Shadow bytes around the buggy address:
  0x0aab2ab25d00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0aab2ab25d10: 00 00 00 00 00 00 00 00 00 00 00 00 f9 f9 f9 f9
  0x0aab2ab25d20: 00 f9 f9 f9 f9 f9 f9 f9 00 00 00 00 00 00 04 f9
  0x0aab2ab25d30: f9 f9 f9 f9 04 f9 f9 f9 f9 f9 f9 f9 04 f9 f9 f9
  0x0aab2ab25d40: f9 f9 f9 f9 04 f9 f9 f9 f9 f9 f9 f9 00 00 00 00
=>0x0aab2ab25d50: 00 00 00 00 00 00 00 00 f9[f9]f9 f9 00 00 00 00
  0x0aab2ab25d60: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0aab2ab25d70: 00 00 00 00 00 00 00 00 00 00 00 00 f9 f9 f9 f9
  0x0aab2ab25d80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0aab2ab25d90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0aab2ab25da0: f9 f9 f9 f9 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==1120727==ABORTING

**Comment by giulianobelinassi** [comment]
The issue seems to be a simple translation error from assembly to C. Where is defined
```
char tmp = s[xrat >> 16];
```
It should be
```
unsigned char tmp = s[xrat >> 16];
```
otherwise, negative access to remap array will be performed, causing a invalid read.

**Comment by giulianobelinassi** [comment]
Fix in #655

**Comment by giulianobelinassi** [comment]
Fixed in cfa1a82df2c098bff2191722fbb3f9c79294d3ef

**Comment by mvdhout1992** [comment]
The function is translated from the original in assembly to C. I think @OmniBlade did the conversion.
